simulator/func_sim/multiplication.h

```diff
@@ -0,0 +1,126 @@
+/*
+* multiplication.h - implementation of riscv multiplication
+* @author Yauheni Sharamed SHaramed.EI@phystech.edu
+* Copyright 2014-2019 MIPT-MIPS
+*/
+
+
+template<typename T>
+auto riscv_multiplication_low(T x, T y) {
+    using UT = unsign_t<T>;
+    auto lo = narrow_cast<UT>( x * y & all_ones<UT>());
+    return lo;
+}
+
+#if defined (__SIZEOF_INT128__)
+// For RISCV-128bit result of multiplication is 256 bit type,
+// which is not defined in ABI.
+// So, we have to use Karatsuba algorithm to get high register of 
+// multiplication result for unsigned*unsigned.
+template<typename T>
```

Comment:
Can these implementations be used with Boost types?
I agree that Boost-specific code is clearer and maybe faster, but supporting two versions of the same code has never been a pleasure...